### ISSUE TITLE
Inconsistent sparse_categorical_crossentropy on cpu vs gpu

### ISSUE THREAD

**Issue by cheyennee** [tensorflow/tensorflow#57173]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

TF2.4

### Custom Code

Yes

### Current Behaviour?

```shell
the result on CPU should be consistent with the result on GPU. However, I find the result on CPU is not equal to the result on GPU on some APIs.
These APIs are:
tf.keras.losses.sparse_categorical_crossentropy
tf.keras.metrics.SparseCategoricalCrossentropy
tf.losses.sparse_categorical_crossentropy
tf.metrics.SparseCategoricalCrossentropy
```

### Standalone code to reproduce the issue

```shell
results = dict()
import tensorflow as tf
try:
  try:
    with tf.device('/CPU'):
      arg_0_0 = -56
      arg_0_1 = -47
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0_0 = 0.05
      arg_1_0_1 = 0.95
      arg_1_0_2 = 0
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1_0 = 0.1
      arg_1_1_1 = 0.8
      arg_1_1_2 = 0.1
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_cpu"] = tf.keras.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_cpu"] = "Error:"+str(e)
  try:
    with tf.device('/GPU:0'):
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_gpu"] = tf.keras.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_gpu"] = "Error:"+str(e)
except Exception as e:
  results["err"] = "Error:"+str(e)
print(results)
```
The results is {'err_cpu': 'Error:Received a label value of -56 which is outside the valid range of [0, 3). Label values: -56 -47 [Op:SparseSoftmaxCrossEntropyWithLogits]', 'res_gpu': <tf.Tensor: shape=(2,), dtype=float32, numpy=array([nan, nan], dtype=float32)>}.

```
results = dict()
import tensorflow as tf
try:
  try:
    with tf.device('/CPU'):
      arg_0_0 = 28
      arg_0_1 = -8
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0_0 = -49.95
      arg_1_0_1 = -3.05
      arg_1_0_2 = False
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1_0 = -2.9
      arg_1_1_1 = -44.2
      arg_1_1_2 = -21.9
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_cpu"] = tf.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_cpu"] = "Error:"+str(e)
  try:
    with tf.device('/GPU:0'):
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_gpu"] = tf.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_gpu"] = "Error:"+str(e)
except Exception as e:
  results["err"] = "Error:"+str(e)
print(results)
```
The result is {'err_cpu': 'Error:Received a label value of -8 which is outside the valid range of [0, 3).  Label values: 28 -8 [Op:SparseSoftmaxCrossEntropyWithLogits]', 'res_gpu': <tf.Tensor: shape=(2,), dtype=float32, numpy=array([nan, nan], dtype=float32)>}.

CPU throws an error, while GPU runs.
```


### Relevant log output

_No response_</details>

**Comment by vulkomilev** [comment]
Hi,
Please the error msg:'err_cpu': 'Error:Received a label value of -8 which is outside the valid range of [0, 3). .Here is a fixed version
`
results = dict()
import tensorflow as tf
try:
  try:
    with tf.device('/CPU'):
      arg_0_0 = 1
      arg_0_1 = 2
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0_0 = 0.05
      arg_1_0_1 = 0.95
      arg_1_0_2 = 0
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1_0 = 0.1
      arg_1_1_1 = 0.8
      arg_1_1_2 = 0.1
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_cpu"] = tf.keras.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_cpu"] = "Error:"+str(e)
  try:
    with tf.device('/GPU:0'):
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_gpu"] = tf.keras.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_gpu"] = "Error:"+str(e)
except Exception as e:
  results["err"] = "Error:"+str(e)
print(results)
`
`results = dict()
import tensorflow as tf
try:
  try:
    with tf.device('/CPU'):
      arg_0_0 = 1
      arg_0_1 = 2
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0_0 = -49.95
      arg_1_0_1 = -3.05
      arg_1_0_2 = False
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1_0 = -2.9
      arg_1_1_1 = -44.2
      arg_1_1_2 = -21.9
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_cpu"] = tf.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_cpu"] = "Error:"+str(e)
  try:
    with tf.device('/GPU:0'):
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_gpu"] = tf.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_gpu"] = "Error:"+str(e)
except Exception as e:
  results["err"] = "Error:"+str(e)
print(results)

`

**Comment by cheyennee** [comment]
> 

I know there is something wrong with arg_0_0 and arg_0_1. But what I care about is that cpu and gpu handle these error values differently. 🥺

**Comment by cheyennee** [comment]
> Hi, Please the error msg:'err_cpu': 'Error:Received a label value of -8 which is outside the valid range of [0, 3). .Here is a fixed version results = dict() import tensorflow as tf try: try: with tf.device('/CPU'): arg_0_0 = 1 arg_0_1 = 2 arg_0 = [arg_0_0,arg_0_1,] arg_1_0_0 = 0.05 arg_1_0_1 = 0.95 arg_1_0_2 = 0 arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,] arg_1_1_0 = 0.1 arg_1_1_1 = 0.8 arg_1_1_2 = 0.1 arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,] arg_1 = [arg_1_0,arg_1_1,] results["res_cpu"] = tf.keras.losses.sparse_categorical_crossentropy(arg_0,arg_1,) except Exception as e: results["err_cpu"] = "Error:"+str(e) try: with tf.device('/GPU:0'): arg_0 = [arg_0_0,arg_0_1,] arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,] arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,] arg_1 = [arg_1_0,arg_1_1,] results["res_gpu"] = tf.keras.losses.sparse_categorical_crossentropy(arg_0,arg_1,) except Exception as e: results["err_gpu"] = "Error:"+str(e) except Exception as e: results["err"] = "Error:"+str(e) print(results)
> 
> results = dict() import tensorflow as tf try: try: with tf.device('/CPU'): arg_0_0 = 1 arg_0_1 = 2 arg_0 = [arg_0_0,arg_0_1,] arg_1_0_0 = -49.95 arg_1_0_1 = -3.05 arg_1_0_2 = False arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,] arg_1_1_0 = -2.9 arg_1_1_1 = -44.2 arg_1_1_2 = -21.9 arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,] arg_1 = [arg_1_0,arg_1_1,] results["res_cpu"] = tf.losses.sparse_categorical_crossentropy(arg_0,arg_1,) except Exception as e: results["err_cpu"] = "Error:"+str(e) try: with tf.device('/GPU:0'): arg_0 = [arg_0_0,arg_0_1,] arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,] arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,] arg_1 = [arg_1_0,arg_1_1,] results["res_gpu"] = tf.losses.sparse_categorical_crossentropy(arg_0,arg_1,) except Exception as e: results["err_gpu"] = "Error:"+str(e) except Exception as e: results["err"] = "Error:"+str(e) print(results)
> 
> tl;dr i have changed these variables arg_0_0 = 1 arg_0_1 = 2

I know there is something wrong with arg_0_0 and arg_0_1. But what I care about is that cpu and gpu handle these error values differently. 🥺

**Comment by sushreebarsa** [comment]
@cheyennee Could you please confirm if this is a duplicate ticket of this [issue](https://github.com/tensorflow/tensorflow/issues/57170).
Thank you!

**Comment by tilakrayal** [comment]
@cheyennee,
TensorFlow 2.4 is not actively supported. Could you please update TensorFlow to the latest stable version v2.9.0 and check if you are facing the same issue. Thank you!

**Comment by cheyennee** [comment]
@tilakrayal 
I repro the code on colab [https://colab.research.google.com/drive/1mwRKRPO6pdPyyi3AGoTBwOaE4e4oJKMp?usp=sharing](#57173). The tensorflow version is 2.8, CPU and GPU output same result. I think the problem may have been fixed. Thanks.🤝
